src/dft.cpp

```diff
@@ -1375,4 +1388,94 @@ void fields::get_mode_mode_overlap(void *mode1_data, void *mode2_data, dft_flux
+  size_t *reduced_dims = reduce_array_dimensions(where, rank, reduced_rank, reduced_grid_size, dims, stride, reduced_stride, dirs, reduced_dirs);
+
+  std::vector<int> monitor_array = {1, 1, 1}; // lengths of the dft monitor along the three dimensions
+  for (int nd = 0; nd < 3; ++nd) if (nd < reduced_rank) monitor_array[nd] = reduced_dims[nd];
+
+  return monitor_array;
+}
+
+std::vector<struct sourcedata> dft_fields::fourier_sourcedata(const volume &where, component c, fields &f, const std::complex<double>* dJ){
+  const size_t Nfreq = freq.size();
+
+  ivec min_corner, max_corner;
+  int rank, reduced_rank;
+  direction dirs[3], reduced_dirs[3];
+  size_t array_size, bufsz, reduced_grid_size, reduced_stride[3], stride[3];
+  dft_chunk *chunklists[1];
+  chunklists[0] = chunks;
+  
+  size_t *dims = f.get_dims(chunklists, 1, c, min_corner, max_corner, array_size, bufsz, rank, dirs);
+  size_t *reduced_dims = reduce_array_dimensions(where, rank, reduced_rank, reduced_grid_size, dims, stride, reduced_stride, dirs, reduced_dirs);
+  size_t monitor_array[3] = {1, 1, 1}; // lengths of the dft monitor along the three dimensions
+  for (int nd = 0; nd < 3; ++nd) if (nd < reduced_rank) monitor_array[nd] = reduced_dims[nd];
+  const size_t monitor_size = monitor_array[0]*monitor_array[1]*monitor_array[2]; // total number of points in the monitor
+
+  std::vector<struct sourcedata> temp;
+
+  for (dft_chunk *f = chunks; f; f = f->next_in_dft) {
+    assert(Nfreq == f->omega.size());
+    vec rshift(f->shift * (0.5 * f->fc->gv.inva));
+
+    std::vector<ptrdiff_t> idx_arr;
+    std::vector<std::complex<double> > amp_arr;
+    std::complex<double> EH0 = std::complex<double>(0,0);
+    sourcedata temp_struct = {component(f->c), idx_arr, f->fc->chunk_idx, amp_arr};
+    int position_array[3] = {0,0,0}; // array indicating the position of a point relative to the minimum corner of the monitor
+
+    LOOP_OVER_IVECS(f->fc->gv, f->is, f->ie, idx) {
+      IVEC_LOOP_LOC(f->fc->gv, x0);
+      IVEC_LOOP_ILOC(f->fc->gv, ix0);
+      x0 = f->S.transform(x0, f->sn) + rshift;
+      ix0 = f->S.transform(ix0, f->sn) + f->shift;
+
+      double dJ_weight = 1; // weight for linear interpolation
+      int nd = 0;
+      LOOP_OVER_DIRECTIONS(f->fc->gv.dim, d){
+        if (where.in_direction(d) > 0) position_array[nd++] = int((ix0.in_direction(d)-min_corner.in_direction(d))/2);
+        else dJ_weight *= (1-abs(x0.in_direction(d)-where.in_direction_min(d))/(f->fc->gv.inva)); // based on distances
+      }
+
+      // index when dJ is flattened to a one-dimenional array
+      size_t idx_1d = position_array[0]*monitor_array[1]*monitor_array[2]+position_array[1]*monitor_array[2]+position_array[2];
+
+      if (f->avg1==0 && f->avg2==0){ // yee_grid = true
+        temp_struct.idx_arr.push_back(idx);
+        for (size_t i = 0; i < Nfreq; ++i) {
+          EH0 = dJ_weight*dJ[monitor_size*i+idx_1d];
+          if (is_electric(temp_struct.near_fd_comp)) EH0 *= -1;
```

Comment:
Why flip sign here?

Reply:
cc @mochen4

Reply:
The flip of sign is consistent with [Owen Miller's thesis](http://optoelectronics.eecs.berkeley.edu/ThesisOwenMiller.pdf) Eq (5.15b). It says the change of permittivity ε in the design region induces an electric dipole, whose effects on E and H in some region χ, is equivalent to change of E field in the design region from the electric and negative magnetic dipole in the region χ, by reciprocity.